graphql_compiler/schema_transformation/utils.py

```diff
@@ -46,16 +46,62 @@ class SchemaStructureError(SchemaTransformError):
-    required by GraphQL), and doesn't start with double underscores as such type names are
+    This may be raised if the input schema contains invalid names. A name is considered valid if it
+    consists of alphanumeric characters and underscores and doesn't start with a numeric character
+    (as required by GraphQL), and doesn't start with double underscores as such type names are
     reserved for GraphQL internal use.
+
+    Not to be confused with SchemaRenameInvalidNameError, which is used for schemas being renamed
+    and includes more information.
     """
 
 
+class SchemaRenameInvalidNameError(SchemaTransformError):
+    """Raised if a type name is not valid during renaming.
+
+    A name is considered valid if it consists of alphanumeric characters and underscores and doesn't
+    start with a numeric character (as required by GraphQL), and doesn't start with double
+    underscores as such type names are reserved for GraphQL internal use.
+
+    Not to be confused with InvalidTypeNameError, which is used for schemas that aren't being
+    renamed because this type of error includes information from the renaming process that
+    InvalidTypeNameError doesn't.
+    """
+
+    # TODO(Leon): Would like feedback on this-- is there a better way to do this than have two types
+    # of invalid name errors? See docstring for why I split them up.
+    invalid_type_names: Dict[str, str]
+
+    def __init__(self, invalid_type_names: Dict[str, str]) -> None:
+        """Record all invalid names that arise specifically from renaming."""
+        if not invalid_type_names:
+            raise ValueError(
+                "Cannot raise SchemaRenameInvalidNameError without at least one invalid name, but "
+                "all arguments were empty dictionaries."
+            )
+        super().__init__()
+        self.invalid_type_names = invalid_type_names
+
+    def __str__(self) -> str:
+        """Explain invalid names and the fix."""
+        explanation = (
+            "Applying the type renaming would rename types with names that are not valid, "
+            "unreserved GraphQL names. Valid, unreserved GraphQL names must consist of only "
+            "alphanumeric characters and underscores, must not start with a numeric character, and "
+            "must not start with double underscores."
+        )
+        sorted_invalid_type_names = sorted(self.invalid_type_names.items())
+        invalid_type_names_message = (
+            f"The following is a list of tuples that describes what needs to be fixed for type "
+            f"renamings. Each tuple is of the form (original_name, invalid_new_name) where "
+            f"original_name is the name in the original schema and invalid_new_name is what "
+            f"original_name would be renamed to: {sorted_invalid_type_names}"
+        )
+        return "\n".join(filter(None, [explanation, invalid_type_names_message]))
```

Comment:
This is also an artifact from splitting code off the large field renamings draft PR in #959-- when field renamings come into play, we'll also have an invalid names argument for fields, which means that there could be situations where one of the invalid names variables ends up being `None` (e.g. when if the only invalid names are type names, not field names). The filter is for that purpose-- but since right now only type renamings are supported in this PR, it doesn't make immediate sense. Will remove